### PR TITLE
feat(nullability): naive coalesce interpretation

### PR DESCRIPTION
**Pull request by janpaepke**
This implements a naive support for `coalesce`.

This MR builds on https://github.com/mmkal/slonik-tools/pull/375, pls review and merge first.
I split the changes into two MRs, because 375 has a lot less changes and can probably be merged without any further discussions.

The proposed feature implementation does not follow up on nested functions, but it will add support to properly type
```sql
select coalesce(sum(col), 0) as sum
```
As
```typescript
sum: number;
```
rather than
```typescript
sum: number | null;
```

As the most common use-case is to fall back to a static value, this will increase usefulness for most cases.
If it doesn't take, it will always fall back to being nullable, which is the current behaviour, so no harm done.

While working on this I had to wrap my head around the inner workings of `column-info.ts` and `parse.ts`.
There are two notes here:
1. I optimised the mappings loop in `parse.ts` to only iterate once, instead of three times. This wasn't so much for the benefit of performance as it was to clarify what's actually happening.
2. It seems the responsibilities of `column-info.ts` and `parse.ts` are somewhat mixed.
`column-info/isNonNullableField` (formerly known as `isFieldNotNull`) has basically the same task as `parse/getAliasMappings`: The extension of field info based on parsing the query

Currently they take completely separate paths, which also means that the sql will actually be parsed not once per query (as you might think) but once per query AND again once per field (called from the above function).

Obviously this is kind of wasteful and confusing.

My suggestion would be to use this flow:
- for each describedQuery retrieved from gdesc
    - parse AST from query
    - foreach statement in AST:
        - run function determining if it is an assignable statement (ref / call / integer) etc and normalises it to include a name and a list of possible tableRefs
        - find field(s) with matching name (/alias) and table in possible tableRefs
            - if multiple fields: ambiguous -> can't retrieve further info
            - if single field: push normalized info into field property
- for each field:
    - check if multiple info is available
    - if yes: something went wrong, we have overlapping aliases or something
    - if no: we're good, process field meta info from normalized properties.
        - now we can decide nullability for all supported statement types (refs/calls/...)
        - add whatever other information we might want to attach to `AnalysedQueryField` (i.e. column? function details?)

### CODE REVIEW

**Comment by mmkal** [comment]
@janpaepke thank you for taking the time to make this change and to review the code as it currently is. I totally agree with you. I mentioned in https://github.com/mmkal/slonik-tools/issues/366#issuecomment-1026843304 that I have tried and failed to make a (small) change which allows for tracking nullability through CTEs, functions, views etc. The messiness you've pointed out is more or less why it was too difficult. 

I'll get #375 merged, then this (I have a couple of minor comments to make). After that would you have any interest in opening a PR which refactors to work something like what you proposed above?

**Comment by mmkal** [comment]
@janpaepke I merged two of your other PRs, and now this is conflicted. Could you fix? 

In the meantime I've published a canary package with this to @slonik/typegen@0.11.4-featureinterpret-coalesce-2022-03-22-16-48-05.6+aeb99ae in case that's helpful.

**Comment by janpaepke** [comment]
oh sure. it conflicted, cause you squash merged the PRs - it "forgot" that it was based on the same commits.
I fixed it and pushed.

> After that would you have any interest in opening a PR which refactors to work something like what you proposed above

Obviously this would be more of an extensive refactor. I do feel I am more familiar with the codebase at this point to potentially take a crack at it. But I want to fix the CROSS Join issue in pgsql-ast-parser first (half way done).
And then there's time. so no promises. ;)

**Comment by mmkal** [comment]
@janpaepke I added a couple of eslint rules (minor stylistic things, just catching yoda expressions and `==` rather than `===`), and a unit test that's failing. Since my team rely on this repo in production and we sometimes go long stretches without looking at it, we're aiming for 100% coverage. In this case, I'm still _slightly_ trying to get my head around the logic. If you pull the branch you can see that the failing test assertion is:

```ts
expect(isNonNullableField('select coalesce(1, count(1)) as x', field)).toBe(true)
```

Is the assertion wrong? I thought from [this comment](https://github.com/janpaepke/slonik-tools/blob/f2e47e7a0f7fd27ca229f1432e1bf59e523141e5/packages/typegen/src/query/column-info.ts#L274-L275) that one level of nesting would be supported.

And last thing - the first assertion checks:

```ts
expect(isNonNullableField('select 1 as x', field)).toBe(false)
```

But really that _could_ be considered non-nullable right?

Anyway, feel free to add to/edit the tests however you'd like - or I can do it if you give me some pointers.

**Comment by janpaepke** [comment]
Hey.
I'll have a look, but generally I use `==` very deliberately, i.e. when comparing nullish. So, when I write `x == null` i mean check if `x` is `null` or `undefined`.
But if you prefer I can make this more explicit with `x === null || x === undefined`.
It's just more verbose.

1st assertion: The assertion is correct, but it's wrong to expect it to be working at this point. Basically it only checks if the last item is cast to something we know to be not-null without any further nesting. This works great for primitives, but if it's a function, it does not use recursion, it gives up. This is the reason, why I called it "naive" in the title, but still think it's useful, as it covers a lot of use cases.
For this to be implemented properly it should be a lot more like I outlined before, as the current implementation is not really set up for recursion and would parse the sql on every call

2nd assertion: Again the assertion is correct, but expecting it to work is not.
This is because of the mentioned parallel paths for determining nullability. Without having tested it, I am pretty sure it would already be determined from the initial parse providing `is_underlying_nullable`. So it doesn't even get to calling `isNonNullableField` for a field like this. This is part of what I was describing what's making this confusing. This also can (and should) be fixed with the refactor.

**Comment by janpaepke** [comment]
Hey Misha,

Hehe, it's really funny because a couple of years ago the roles were reversed and I was on your side of this exact discussion.
A very smart friend of mine convinced me to reconsider, after I reviewed code of him and demanded strict equal check.
His point was: "In contrast to popular belief in the js community, thruthiness and nullishness are very different things."
The same way that `a = b || c` means something very different than `a = b ?? c`.
Since then I have paid a lot more attention to what exactly it is I want to check for and it has made my code better and more reliable.

Interestingly enough that same friend also uses "literals/constants first" in comparisons (`if ('cast' === arg.type)`), which looked very weird to me in the beginning and I seem to have subconsciously adopted.

Both points (the second even more so) come down to preference, so I have no problem changing them.

File: packages/typegen/src/query/column-info.ts [hunk 1]
```diff
@@ -232,21 +235,54 @@ export const defaultAnalysedQueryField = (f: QueryField): AnalysedQueryField =>
...
+    if (c.expr.function.name === 'count') {
+      // `count` is the only aggregation function, which never returns null.
+      return true
+    }
+    if (c.expr.function.name === 'coalesce') {
+      // let's try to check the args for nullability - starting at the end.
+      return c.expr.args
+        .slice(1)
+        .reverse()
+        .find(arg => {
```
[the remark this answers is not in the document]
Comment (janpaepke) [reply packages/typegen/src/query/column-info.ts:273]: yes, it should, good find.

```suggestion
        .some(arg => {
```

This probably means we should add another test like `coalesce(sum(1), 1, null)`, which we'd expect to be nullable.

[the remark this answers is not in the document]
Comment (mmkal) [reply packages/typegen/src/query/column-info.ts:273]: Hmm not sure why it would make a difference. Given every `arg` should always be truthy, I'd have thought `.some` would return a boolean and `.find` would return a truthy object. The result is being tested for truthiness so the only reason I suggested the change was to make the intention clearer, not because it affects functionality. Having said that, that sounds like a sensible test to include either way.

[the remark this answers is not in the document]
Comment (janpaepke) [reply packages/typegen/src/query/column-info.ts:273]: Well this change is actually not only cosmetic, it's also faster.

`find` will iterate through the whole array regardless, while `some` will stop at the first `true`-returning callback.

You might argue that the difference is negligible here and that might be so for most cases (typically we won't have 100s of args), but this is precisely the reason why I rarely rely on `truthy`, as explained above.

This is how apps get slow and subtle bugs creep in.

**Comment by janpaepke** [comment]
So this was a really good discussion, as it pointed me to a thinking error on my part.
In contrast to what I said [here](https://github.com/mmkal/slonik-tools/pull/376#discussion_r841108568), `coalesce(sum(1), 1, null)` should actually non-nullable. The last `null` is ignored, because `coalesce` will return the first non-null value, which of course is the whole point.

Similarly the fairly pointless, but nontheless valid query `select coalesce(1)` should also be non-nullable.

The good new is this makes the code more straightforward.
I optimised the code, improved the comments and added three more tests.

Much better now.

File: packages/typegen/src/query/column-info.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmkal) [reply packages/typegen/src/query/column-info.ts]: `find` stops at the first truthy-returning callback too. You can see this in a node repl:

```
> [1,2,3,4,5,6,7].find(n => { console.log({ n }); return n > 3 })
{ n: 1 }
{ n: 2 }
{ n: 3 }
{ n: 4 }
4

> [1,2,3,4,5,6,7].some(n => { console.log({ n }); return n > 3 })
{ n: 1 }
{ n: 2 }
{ n: 3 }
{ n: 4 }
true
```

Note that neither goes through the whole array. And I still like truthiness checks (when TypeScript can prevent the gotchas). Either way, I like it well enough for cosmetic reasons so it doesn't really matter!

[the remark this answers is not in the document]
Comment (janpaepke) [reply packages/typegen/src/query/column-info.ts]: yes, you are right again. where is my mind? :D

let's get this shipped.

**Comment by mmkal** [comment]
Thanks again @janpaepke!
